generator/import.go

```diff
@@ -1,49 +1,54 @@
 package generator
 
 import (
-	"log"
-	"sort"
+	"go/types"
 	"strings"
 )
 
+// Imports indexes imports by package path and alias so that all imports have a
+// unique alias, and no package is included twice.
+type Imports struct {
+	ByAlias   map[string]Import
+	ByPkgPath map[string]Import
+}
+
+func newImports() Imports {
+	return Imports{
+		ByAlias:   make(map[string]Import),
+		ByPkgPath: make(map[string]Import),
+	}
+}
+
 // Import is a package import with the associated alias for that package.
 type Import struct {
-	Alias string
-	Path  string
+	Alias   string
+	PkgPath string
 }
 
 // AddImport creates an import with the given alias and path, and adds it to
 // Fake.Imports.
-func (f *Fake) AddImport(alias string, path string) Import {
+func (i *Imports) Add(alias string, path string) Import {
+	// TODO: why is there extra whitespace on these args?
 	path = unvendor(strings.TrimSpace(path))
 	alias = strings.TrimSpace(alias)
-	for i := range f.Imports {
-		if f.Imports[i].Path == path {
-			return f.Imports[i]
-		}
+
+	imp, exists := i.ByPkgPath[path]
+	if exists {
+		return imp
 	}
-	log.Printf("Adding import: %s > %s\n", alias, path)
-	result := Import{
-		Alias: alias,
-		Path:  path,
+
+	imp, exists = i.ByAlias[alias]
+	if exists {
+		alias = uniqueAliasForImport(alias, i.ByAlias)
 	}
-	f.Imports = append(f.Imports, result)
-	return result
-}
 
-// SortImports sorts imports alphabetically.
-func (f *Fake) sortImports() {
-	sort.SliceStable(f.Imports, func(i, j int) bool {
-		if f.Imports[i].Path == "sync" {
-			return true
-		}
-		if f.Imports[j].Path == "sync" {
-			return false
-		}
-		return f.Imports[i].Path < f.Imports[j].Path
-	})
+	result := Import{Alias: alias, PkgPath: path}
+	i.ByPkgPath[path] = result
+	i.ByAlias[alias] = result
+	return result
 }
 
+// TODO: what version of go required this?
```

Comment:
This is required in go 1.11 and go 1.10 when using `go/packages` to load packages. The files that are loaded from vendor folders will include the vendor path in the package path, making it invalid to use as an import path unless we strip out the vendor prefix.

This code is functionally identical to code used in golang.org/x/tools packages.